src/lib/homestore_backend/hs_shard_manager.cpp

```diff
@@ -153,10 +153,26 @@ void HSHomeObject::on_shard_message_commit(int64_t lsn, sisl::blob const& header
     }
 
     // hs_ctx will be nullptr when HS is restarting and replay all commited log entries from the last checkpoint;
-    // but do we really need to handle this for create_shard or seal_shard?
-    // because this raft log had been commit before HO restarts and commit result is already saved in HS metablks
-    // when HS restarts, all PG/Shard infos will be recovered from HS metablks and if we commit again, it will cause
-    // duplication;
+    // most of time, the create_shard/seal_shard on_commit() is already completed before restarting/crash and shard info
+    // is already be written into metablk, so we need to do nothing in this case to avoid duplication when replay this
+    // journal log. but there is still a smaller chance that HO is stopped/crashed before writing metablk is called or
+    // completed and we need to recover shard info from journal log in such case.
+    sisl::sg_list value;
+    value.size = blkids.blk_count() * repl_dev->get_blk_size();
+    auto value_buf = iomanager.iobuf_alloc(512, value.size);
+    value.iovs.push_back(iovec{.iov_base = value_buf, .iov_len = value.size});
+    // header will be released when this function returns, but we still need the header when async_read() finished.
+    auto header_ptr = r_cast< const ReplicationMessageHeader* >(header.bytes);
+    repl_dev->async_read(blkids, value, value.size)
```

Comment:
actually, when restarting,  we want to call `do_shard_message_commit` for the specific shard. But the shard info is in `value` which has been written as a block, so we need to read the block and get the value. this is what here does, right?

i go through the solo_dev, it will first write the block, and then write the log.
https://github.com/eBay/HomeStore/blob/0f460de461b64232d03c235ccfbff97274a0bc60/src/lib/replication/repl_dev/solo_repl_dev.cpp#L45

so if crash happens after the block is written but before the log is written, when restarting , no log for this shard exists, we can not find this issue, and a stale shard info block will be left in the chunk. (not sure will gc handle this?)

if we write log first, then when restart , maybe we cannot read `value` from the disk(the same case as above).
 
so , maybe we should think about what we should do here later?

Reply:
both mentioned cases is not this PR want to resolve because if the log is not written successfully, the created shard must be failed and regards nothing happens(but may leave a stale data blk and need GC to recycle it). 
what the PR wants to do is when log is already been written, but crashed before on_commit() called or  on_commit() is called but shard metablk is not written yet.  As the log is persisted, so data must be persisted before and we will replay the log to restore the shard info.

Reply:
we can record this case, and no need to solve in this PR. let`s  revisit this issue after raft_repl_dev is merged.
other parts looks good to me!